File: AEDS2/estruturas_Abstratas_TAD/pilha/pilhaCabeca.java
package AEDS.AEDS2.estruturas_Abstratas_TAD.pilha;

class No {
    int dado;
    No prox;
    No(int d) { dado = d; prox = null; }
}

public class pilhaCabeca {
    No cabeca; // nó cabeça

    pilhaCabeca() {
        cabeca = new No(0);
    }

    void push(int valor) {
        No novo = new No(valor);
        novo.prox = cabeca.prox;
        cabeca.prox = novo;
    }

    int pop() {
        if (cabeca.prox == null) {
            System.out.println("Pilha vazia!");
            return -1;
        }
        No removido = cabeca.prox;
        cabeca.prox = removido.prox;
        return removido.dado;
    }

    void mostrar() {
        No aux = cabeca.prox;
        System.out.print("Topo -> ");
        while (aux != null) {
            System.out.print(aux.dado + " ");
            aux = aux.prox;
        }
        System.out.println();
    }

    public static void main(String[] args) {
        pilhaCabeca p = new pilhaCabeca();
        p.push(10);
        p.push(20);
        p.push(30);
        p.mostrar();
        p.pop();
        p.mostrar();
    }
}

File: AEDS2/estruturas_Abstratas_TAD/pilha/pilhaCabeca.c
#include <stdio.h>
#include <stdlib.h>

typedef struct No {
    int dado;
    struct No *prox;
} No;

typedef struct {
    No *cabeca; // nó cabeça
} Pilha;

void inicializar(Pilha *P) {
    P->cabeca = (No*) malloc(sizeof(No));
    P->cabeca->prox = NULL;
}

void push(Pilha *P, int valor) {
    No *novo = (No*) malloc(sizeof(No));
    novo->dado = valor;
    novo->prox = P->cabeca->prox;
    P->cabeca->prox = novo;
}

int pop(Pilha *P) {
    if (P->cabeca->prox == NULL) {
        printf("Pilha vazia!\n");
        return -1;
    }
    No *removido = P->cabeca->prox;
    int valor = removido->dado;
    P->cabeca->prox = removido->prox;
    free(removido);
    return valor;
}

void mostrar(Pilha *P) {
    No *aux = P->cabeca->prox;
    printf("Topo -> ");
    while (aux != NULL) {
        printf("%d ", aux->dado);
        aux = aux->prox;
    }
    printf("\n");
}

int main() {
    Pilha P;
    inicializar(&P);
    push(&P, 10);
    push(&P, 20);
    push(&P, 30);
    mostrar(&P);
    pop(&P);
    mostrar(&P);
    return 0;
}